letsencrypt-win-simple/Scripts/ImportExchange.ps1

```diff
@@ -1,56 +1,95 @@
 ## example: $ExchangeServices = IMAP,POP,IIS
 ## That will import the certificate for IMAP, POP, and IIS roles in exchange
 ## Valid services:  IMAP | POP | UM | IIS | SMTP | Federation | UMCallRouter
 
-## THIS SCRIPT IS INCOMPLETE AND UNTESTED
+## THIS SCRIPT IS INCOMPLETE AND *mostly* UNTESTED (some modifications have come in from people using it successfully)
 ## Documentation referenced from https://technet.microsoft.com/en-us/library/aa997231(v=exchg.160).aspx
 
 # Should work with exchange 2007 and higher
-Get-PSSnapin -registered | ? {$_.Name -match "Microsoft.Exchange.Management.PowerShell" -and ($_.Name -match "Admin" -or $_.Name -match "E2010" -or $_.Name -match "SnapIn")} | Add-PSSnapin -ErrorAction SilentlyContinue
+Get-PSSnapin -registered | Where-Object {$_.Name -match "Microsoft.Exchange.Management.PowerShell" -and ($_.Name -match "Admin" -or $_.Name -match "E2010" -or $_.Name -match "SnapIn")} | Add-PSSnapin -ErrorAction SilentlyContinue
 
 #$OldThumbprint = (Get-Item -Path RDS:\GatewayServer\SSLCertificate\Thumbprint).CurrentValue
 $CertInStore = Get-ChildItem -Path Cert:\LocalMachine -Recurse | Where-Object {$_.thumbprint -eq $NewCertThumbprint} | Sort-Object -Descending | Select-Object -f 1
-if($CertInStore){
-    try{
-        # Cert must exist in the personal store of machine to bind to RD Gateway
-        if($CertInStore.PSPath -notlike "*LocalMachine\My\*"){
-            $SourceStoreScope = 'LocalMachine'
-            $SourceStorename = $CertInStore.PSParentPath.split("\")[-1]
+try{
+    if($CertInStore.PSPath -notlike "*LocalMachine\My\*"){
+        "Cert thumbprint not found in the cert store... which means we should load it. This means TargetHost and StorePath must be specified"
+        
+        "Try to load certificate from store"
+        $importExchangeCertificateParameters = @{
+        FileName = (Join-Path -Path $StorePath -ChildPath "$TargetHost.pfx")
+        FriendlyName = $TargetHost
+        PrivateKeyExportable = $true
+        }
+        
+        $null = Import-ExchangeCertificate @importExchangeCertificateParameters -ErrorAction Stop
 
-            $SourceStore = New-Object  -TypeName System.Security.Cryptography.X509Certificates.X509Store  -ArgumentList $SourceStorename, $SourceStoreScope
-            $SourceStore.Open([System.Security.Cryptography.X509Certificates.OpenFlags]::ReadOnly)
 
-            $cert = $SourceStore.Certificates | Where-Object {$_.thumbprint -eq $CertInStore.Thumbprint}
+        # Now try to find it again...
+        $CertInStore = Get-ChildItem -Path Cert:\LocalMachine -Recurse | Where-Object {$_.thumbprint -eq $NewCertThumbprint} | Sort-Object -Descending | Select-Object -f 1
+        
+    }
 
+    # Make sure variable is defined
+    $null = Get-ChildItem $CertInStore.PSPath -ErrorAction Stop
+    # Cert must exist in the personal store of machine to bind to exchange services
+    if($CertInStore.PSPath -notlike "*LocalMachine\My\*"){
```

Comment:
Question: Are we using this code path as a fail safe if the `Import-ExchangeCertificate` did not add the certificate to the `LocalMachine\My` store? Is there a scenario when `Import-ExchangeCertificate` does not import the certificate to the `LocalMachine\My` store?

Reply:
Yes, keep in mind the script's original intent was not to be used with centralssl (the fqdn and storepath aren't required parameters)